PK: remove `pk_context::pk_ctx`

valeriosetti opened the pull request:
## Description

Resolves #489 
Might resolve #491

## PR checklist

- [x] **changelog** not required because: "pk_ctx" was a private element in PK structure and `mbedtls_pk_[rsa|ec]` were internal so none of these elements was accessible to the end user.
- [x] **framework PR** not required
- [x] **mbedtls development PR** not required because: no change there
- [x] **mbedtls 3.6 PR** not required because: no backport
- **tests**  not required because: removed code was already unused

valeriosetti committed:
drivers: pk: remove ctx_[alloc|free]_func() from PK context

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
include: pk_private: remove mbedtls_pk_[rsa|ec]

Both functions were used in the past to access "pk_ctx" element of the PK
structure. Since this element is no more used these helper functions are
no more required.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
programs: fuzz: update fuzz_[priv|pub]key

Update both executables as follows:
- remove mbedtls_pk_[rsa|ec] replacing them with a key export and a check
  against the key that was previously parsed;
- remove legacy constructs (ex: entropy, ctr-drbg, ...) which are no more
  needed.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
include|drivers: pk: remove pk_ctx field from PK structure

This field is now absolete and it can be cancelled.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg reviewed (changes requested):
I'm happy with the changes in the library, but there are issues with the fuzzing programs.

Also, there are still two occurrences of `mbedtls_pk_rsa()` in `test_suite_pk.function`. These should be removed, but perhaps more importantly I'd really like to understand how we could get a green CI with those calls present. They must be in a portion of code that's always removed by the pre-processor (otherwise the build would fail with a linke error in some configs) and that might be the indication of a wider problem. Can you look into it?

valeriosetti committed:
programs: fuzz: use static buffers instead of dynamic ones

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on programs/fuzz/fuzz_privkey.c:
```diff
@@ -2,102 +2,64 @@
 
 #include <stdint.h>
 #include <stdlib.h>
+#include <stdio.h>
```
TBH I don't remember exactly why I had to add it... Let's try to remove it and see what the CI thinks. Perhaps it's just a temporary leftover.

mpg replied:
Yes. Should the fuzzing program really print things out? None of the other fuzzing functions do (as opposed to `fuzz_onefile.c` and `fuzz_common.c`).

Basically, as a general rule of thumb, since the fuzzing programs are not the primary goal of this PR, I'd like to minimize changes there, unless justified by a recent change in the library:
- Removing use of `pk_rsa()` is clearly justified by a recent change in the library (this function is no longer public) and necessary for the primary goal of this PR (entirely removing this function).
- Removing `ctr_drbg()` and so is not necessary for this PR, but justified by a recent change in the library, and something we wanted to do anyway (https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/491) so why not.
- Adding `fprintf()`: can't see a reason to make that change now.

I don't think it's wrong, but as I don't see a reason why we should change this now, I'd rather not change it. (In order to reduce noise, and minimise the risk or mistake - which probably doesn't apply here, but for `abort()` it does.)

(Still minor and not a blocker, just trying to explain my reasoning.)

mpg commented on programs/fuzz/fuzz_privkey.c:
```diff
@@ -2,102 +2,64 @@
     if (status != PSA_SUCCESS) {
-        goto exit;
+        fprintf(stderr, "Error: failed to initialize PSA\n");
+        abort();
```
To expand on Gilles's "please don't remove any `abort` calls" -> other that those you introduced in this PR which do not indicate that the input is a "bad" case (ie, a case that makes our parser misbehave). So, essentially just preserve the existing in terms of "abort vs goto exit".

valeriosetti committed:
programs: fuzz: remove printf() calls

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
programs: fuzz: remove memcmp() check

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
programs: fuzz: re-add "goto error" if psa_crypto_init fails

This reset the fuzz_[priv|pub]key behavior back to its original state.
Similarly abort() calls are kept in case key parsing or writing fails.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
programs: fuzz: adjust fuzz_[priv|pub]key guards

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
drivers: pk: fix comment mentionig mbedtls_pk_rsa()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pk: remove remaining occurrences of mbedtls_pk_rsa() from pk_psa_sign()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
tests: pk: remove PK_PARSE_C dependency from pk_psa_sign()

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg reviewed (changes requested):
Looks pretty good to me except I'm still not happy with some changes in the fuzzing programs, which IMO are both unrelated to the purpose of this PR, and wrong.

valeriosetti committed:
programs: fuzz: restore original behavior in case of failures

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented:
> > I wonder it we should also have a component building with `PK_PARSE_C && !PK_WRITE_C`?
> 
> Turns out we already have two: `component_test_config_suite_b` and `component_test_config_thread` but they're both ECC-only, so they won't "see" issues with RSA.

Just to be sure, are you suggesting to add companion components of `component_full_no_pkparse_pkwrite` in `components-configuration-crypto.sh` in Mbed TLS repo?

mpg commented:
> Just to be sure, are you suggesting to add companion components of `component_full_no_pkparse_pkwrite` in `components-configuration-crypto.sh` in Mbed TLS repo?

Not necessarily, I mostly wanted to make sure I understood the current state of our testing.

That said, it might be interesting to add components "full minus pkwrite" and "full minus pkparse" but if we do so, let's add them directly on the correct side, so `tests/scripts/components-configuration.sh` in the crypto repo. And might as well move the related existing component there while at it. But that would be a separate PR.

valeriosetti commented:
> I'm still leaning on having those components only on the crypto side, but we probably want to fix the dependencies in `mbedtls_check_config.h` though.

OK, so here's the proposal then: let's add new PK components to both repos. On the crypto side we'll test all the combinations of parse/write, whereas on the mbedtls one we can limit to only what really makes sense. Plus on the mbedtls side I will also fix the checks in `mbedtls_check_config.h`. Wdyt?

Should I add the new tests components for crypto in this PR or as follow-up? I suggest the latter in order to keep the scope reduced and having faster/easier reviews

Edit: I created the tf-psa-crypto [issue](https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/525). Can I add it to the "1.0 Follow-up" EPIC?

mpg commented:
> Edit: I created the tf-psa-crypto [issue](https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/525). Can I add it to the "1.0 Follow-up" EPIC?

Thanks! Yes, I'm adding it to the EPIC, since it's related to PK which is the main topic.

mpg commented on programs/fuzz/fuzz_privkey.c:
```diff
@@ -2,102 +2,61 @@
-#endif
-#if defined(MBEDTLS_ECP_C)
-        if (mbedtls_pk_get_type(&pk) == MBEDTLS_PK_ECKEY ||
+        if (mbedtls_pk_get_type(&pk) == MBEDTLS_PK_RSA ||
```
While at it, let's stop using `pk_get_type()` which is on its way out. This used to be necessary to differentiate between RSA and EC but we no longer need to since we're handling both in the same way. Also, when key types are added in the future (ML-DSA) we'll want to do the same. Basically any key we've parsed successfully we should be able to also write out successfully, or this is a bug, regardless of the key type.

(Same for the other program of course.)

mpg commented on programs/fuzz/fuzz_privkey.c:
```diff
@@ -2,102 +2,61 @@
-#endif
-        {
+        } else {
             /* The key is valid but is not of a supported type.
```
This is not future-proof when we add support for new key types. Also, completely unsupported types would result in `mbedtls_pk_write_key_der()` returning an error anyway.

mpg reviewed (changes requested):
Almost there! Just requesting one more change in the fuzzing programs to reap all the benefits of the new sanity check.

valeriosetti committed:
programs: fuzz: remove PK type checks and simplify code

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

mpg reviewed (approved):
LGTM, thanks!
